### PR TITLE
Implement liquidation of undercollateralized debt positions

### PR DESCRIPTION
**Pull request by xJonathanLEI**
This PR implements permissionless liquidation of undercollateralized debt positions by enabling liquidators to burn their liquids to pay off other users' debt, and redeem the underlying collateral in the process.

A new contract `LnLiquidation` has been added. Changes have also been made to the following contracts:

- `LnBuildBurnSystem`
- `LnCollateralSystem`
- `LnRewardLocker`

which require an upgrade.

### CODE REVIEW

File: contracts/LnLiquidation.sol [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+        lnBuildBurnSystem.burnForLiquidation(user, msg.sender, lusdToBurn);
+
+        LiquidationRewardCalculationResult memory rewards =
+            calculateRewards(lusdToBurn, evalResult.collateralPrice, ratios.markerRewardRatio, ratios.liquidatorRewardRatio);
+
+        {
+            uint256 totalCollateralToMove = rewards.collateralWithdrawalAmount.add(rewards.totalReward);
+            uint256 totalCollateralAmount = evalResult.stakedCollateral.add(evalResult.lockedCollateral);
+            require(totalCollateralToMove > 0, "LnLiquidation: no collateral withdrawal");
+            require(totalCollateralToMove <= totalCollateralAmount, "LnLiquidation: insufficient collateral"); // Insurance fund needed to resolve this
```
[the remark this answers is not in the document]
Comment (xJonathanLEI) [reply contracts/LnLiquidation.sol:194]: No it simply means we cannot liquidate it in full.

File: contracts/LnCollateralSystem.sol [hunk 2]
```diff
@@ -423,9 +435,21 @@ contract LnCollateralSystem is LnAdminUpgradeable, PausableUpgradeable, LnAddres
...
+        address fromUser,
+        address toUser,
+        bytes32 currency,
+        uint256 amount
+    ) external whenNotPaused onlyLiquidation {
+        userCollateralData[fromUser][currency].collateral = userCollateralData[fromUser][currency].collateral.sub(amount);
+        userCollateralData[toUser][currency].collateral = userCollateralData[toUser][currency].collateral.add(amount);
+        emit CollateralMoved(fromUser, toUser, currency, amount);
+    }
+
```
[the remark this answers is not in the document]
Comment (xJonathanLEI) [reply contracts/LnCollateralSystem.sol:448]: Check what?

File: tests/integration/Liquidation.spec.ts [hunk 3]
```diff
@@ -0,0 +1,591 @@
...
+      stack.lnLiquidation
+        .connect(alice)
+        .removeUndercollateralizationMark(alice.address)
+    ).to.be.revertedWith("LnLiquidation: still undercollateralized");
+
+    // LINA price goes to $0.1. Alice can now remove mark
+    await setLinaPrice(0.1);
+    await stack.lnLiquidation
+      .connect(alice)
+      .removeUndercollateralizationMark(alice.address);
```
[the remark this answers is not in the document]
Comment (xJonathanLEI) [reply tests/integration/Liquidation.spec.ts:226]: Please submit a PR if you wanna add more tests.

File: contracts/LnLiquidation.sol [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+        uint256 stakedCollateral;
+        uint256 lockedCollateral;
+    }
+
+    ILnBuildBurnSystem public lnBuildBurnSystem;
+    ILnCollateralSystem public lnCollateralSystem;
+    ILnConfig public lnConfig;
+    ILnDebtSystem public lnDebtSystem;
+    ILnPrices public lnPrices;
+    ILnRewardLocker public lnRewardLocker;
```
[the remark this answers is not in the document]
Comment (xJonathanLEI) [reply contracts/LnLiquidation.sol:78]: IMO the benefit of setting them private is too marginal to be considered (very, very minimal gas savings).

File: contracts/LnCollateralSystem.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (xJonathanLEI) [reply contracts/LnCollateralSystem.sol:448]: What do you wanna check? Mind posting example code in comments?

File: contracts/LnLiquidation.sol (hunk 1 above)
[the remark this answers is not in the document]
Comment (xJonathanLEI) [reply contracts/LnLiquidation.sol:194]: This situation is a common problem in margin trading (bankruptcy). As specified in the comment insurance fund needs to be implemented to resolve it.

File: contracts/LnCollateralSystem.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (eric-linear) [reply contracts/LnCollateralSystem.sol:448]: I am wondering if we need to add, although Liquidator has checked the collateral amount before in Liquidator contract.
`require(amount <= userCollateralData[fromUser][currency].collateral, "Can not redeem more than collateral");`

[the remark this answers is not in the document]
Comment (xJonathanLEI) [reply contracts/LnCollateralSystem.sol:448]: `.sub()` is `SafeMath`

File: tests/integration/Liquidation.spec.ts [hunk 5]
```diff
@@ -0,0 +1,591 @@
...
+        .markPositionAsUndercollateralized(alice.address)
+    ).to.be.revertedWith("LnLiquidation: not undercollateralized");
+
+    // Price of LINA drops such that Alice's C-ratio falls below liquidation ratio
+    await setLinaPrice(0.038);
+
+    // Can mark position normally
+    await stack.lnLiquidation
+      .connect(bob)
+      .markPositionAsUndercollateralized(alice.address);
```
[the remark this answers is not in the document]
Comment (xJonathanLEI) [reply tests/integration/Liquidation.spec.ts]: Assertion added.

**Pull request merged**